runtime/vm/guardedstorage.c

```diff
@@ -60,13 +69,22 @@ j9gs_initializeThread(J9VMThread *vmThread)
 			gsControlBlock->designationRegister = 0;
 			gsControlBlock->sectionMask = 0;
 			gsControlBlock->paramListAddr = (uint64_t) &vmThread->gsParameters;
-			vmThread->gsParameters.handlerAddr = (uint64_t)(uintptr_t)J9_BUILDER_SYMBOL(handleGuardedStorageEvent);
-		
+			if (1 == supportsGuardedStorageFacility) {
+				vmThread->gsParameters.handlerAddr = (uint64_t)(uintptr_t)J9_BUILDER_SYMBOL(handleGuardedStorageEvent);
+			} else {
```

Comment:
As discussed  we need an extra flag/option to check if we want to really use S/W read barrier or ignore it. Not sure what the option would be .... could be either -Xjit or -Xgc option.

Reply:
Perhaps some thing like -XXgc:emulateGuardedStorage.  I feel it's a Port Library knob/option, but it's just simpler to make it a GC option. The flag itself could be part of the global GS structure, but again, it's probably simpler to put it in the global GC structure (GCExtensions).

-XX (double X) indicates it'd be a private/undocumented option.

Reply:
I'm not an expert on this, but the feature would seem to run functionally correctly with software checks, it is just the performance overhead that would prevent us from using it in the 'real-world'. If the checks could be made cheaper then the software version could work. Perhaps rather than calling the option emulateGuardedStorage something like softGuardedStorage or something like that since it is doing guarded storage from what I understand, just not with the hardware instruction and trap that makes it performant?

Reply:
Thanks for the comment, 

Indeed, emulateGuardedStorage  might not  be the right name - we are not fully emulating the capability of GS (to monitor 64 adjacent fixed sized regions). So, even GS part of this name is not quite right.

We are effectively doing software variant of Scavenger Evacuate range check, for which GS is used if available. 

Internally, we could introduce a new type of write barrier - beside the existing OMR_GC_READ_BARRIER_TYPE_EVACUATE to have OMR_GC_READ_BARRIER_TYPE_EVACUATE_SOFTWARE, but then many places in code (JIT/VM/GC) would need to change, to accept both types, while GC and VM would not really do anything different between the two types. So I want to avoid introducing the new type, but just introduce a flag that only JIT would care about, to know how to act if barrier is 'evacuate' - to generate GS instruction or s/w variant of the barrier.  Based on this, my next candidate for the name is then: -XXgc:softwareEvacuateReadBarrier

Reply:
Added -XXgc:softwareEvacuateReadBarrier option as per the discussion

Reply:
I can't settle on this (seemingly minor detail). Evacuate is actually an action that read barrier performs. What we actually do here in s/w is a trigger, not an action. This is why I started with emulateGuardedStorage (but again, we don't fully emulate it). The trigger we do in s/w is a simplified guarded storage functionality that could be qualified as 'range check'.
Based on it. it should be -XXgc:softRangeCheckReadBarrier.

But unless people really like this newest one, I'm fine to leaving as is.